### PR TITLE
CIP to introduce authored blobs

### PR DESCRIPTION
**Pull request by cmwaters**
This CIP is in reference to this issue: https://github.com/celestiaorg/celestia-app/issues/3316

Note that there is a diagram so when changing the name of the file and folders, make sure to change the name of the folder in the assets directory and update this CIP that references it

### CODE REVIEW

File: cips/cip-blob-verified-signer.md
Comment (evan-forbes) [inline cips/cip-blob-verified-signer.md:33]: > (In zero-knowledge)

From my perspective, I think I know what is meant by "zero-knowledge", but it might be valuable to be more specific or use different wording. I've heard the term "in circuit" used before for constraints applied to the public and private inputs, do you think that applies here or perhaps something else?

**Review by evan-forbes — COMMENTED** [review]
still chewing but otherwise LGTM!

File: cips/cip-blob-verified-signer.md
Comment (preston-evans98) [inline cips/cip-blob-verified-signer.md:14]: nit: This is useful for rollups that don't enshrine a sequencer as well! In fact, we don't enshrine a particular sequencer in Sovereign SDK

File: cips/cip-blob-verified-signer.md
Comment (preston-evans98) [inline cips/cip-blob-verified-signer.md:60]: If the `signer` is empty but a user specifies `share_version = 2`, is the version in the `Info Byte` also changed to be `v1`?

File: cips/cip-blob-verified-signer.md
Comment (preston-evans98) [inline cips/cip-blob-verified-signer.md:70]:
```suggestion
- Verify that the `signer` in each blob matches that of an allowed sequencer
```

File: cips/cip-blob-verified-signer.md
Comment (cmwaters) [reply cips/cip-blob-verified-signer.md:14]: Can you explain that a little further. My mental model was that the nodes had a list of accepted "authors" of blobs. Or perhaps another way of putting it is If anyone can submit blobs then why do you need to verify the signer.

File: cips/cip-blob-verified-signer.md
Comment (preston-evans98) [inline cips/cip-blob-verified-signer.md:72]: This is a good call out. If Celestia also added an RPC endpoint for fetching blob metadata, this could be made quite efficient. Rollups would just have to...
1. Fetch the metadata via RPC
2. For each blob from an allowed sender, fetch the full blob and its proof via RPC
3. (zk-rollups only) For each other blob, fetch *only the first share* of the blob and its proof via RPC to check that the metadata is correct

File: cips/cip-blob-verified-signer.md
Comment (cmwaters) [reply cips/cip-blob-verified-signer.md:33]: I was taking this from a slack thread. I can rephrase to using (in circuit) if you think that would be better

File: cips/cip-blob-verified-signer.md
Comment (cmwaters) [reply cips/cip-blob-verified-signer.md:60]: As I see it (and feel free to correct me if you feel there's a better configuration), a tx specifying a share version of 2 and an empty signer field will be rejected. A tx specifying a share version of 1 with a signer field will also be rejected.

File: cips/cip-blob-verified-signer.md
Comment (cmwaters) [reply cips/cip-blob-verified-signer.md:72]: Iiuc, you mean that there would be a RPC query to fetch blobs by namespace and signer?

File: cips/cip-blob-verified-signer.md
Comment (preston-evans98) [reply cips/cip-blob-verified-signer.md:14]: We have a list of accepted authors, but it's dynamic (no particular enshrined sequencer), and we even accept blobs from people who aren't on the list under some circumstances (i.e. to enable recovery if all sequences are offline/malicious)

Comment (preston-evans98) [reply cips/cip-blob-verified-signer.md:14]: The actual rules are a little bit involved, but the high level idea is that we accept...
- All blobs from registered sequencers (i.e. actors who are subject to slashing on the rollup)
- Up to a (small) constant number of additional blobs per celestia block from unregistered sequencers

File: cips/cip-blob-verified-signer.md
Comment (preston-evans98) [reply cips/cip-blob-verified-signer.md:60]: That sounds great. Can we add a sentence making it explicit in the CIP?

File: cips/cip-blob-verified-signer.md
Comment (cmwaters) [reply cips/cip-blob-verified-signer.md:14]: I see so it's somewhat a middle ground between complete permissionless rollups and enshrined (whitelisted). I'll update the abstract accordingly

File: cips/cip-blob-verified-signer.md
Comment (cmwaters) [reply cips/cip-blob-verified-signer.md:72]: So another way of framing this would be the minimal set of data for the fork-choice rule to filter out what blobs are worth processing or not. Wouldn't this data need to be provable? Wouldn't this make it look similar to the `PFBNamespace` although a little more compact

**Comment by nashqueue** [comment]
Maybe we should add the fact that you need versioning for the CreateCommitment function. As we might be adding other things in the future which will change the shares and therefore the share commitment. 

Create commitment will now have to be aware of the signer.

File: cips/cip-blob-verified-signer.md
Comment (cemozerr) [reply cips/cip-blob-verified-signer.md:14]: > Up to a (small) constant number of additional blobs per celestia block from unregistered sequencers

The constant is so small that effectively rollup should be completely permissionless.

File: cips/cip-blob-verified-signer.md
Comment (rootulp) [inline cips/cip-blob-verified-signer.md:69]: [nit] can this be clarified b/c I'm not sure what it means. Does this step mean:

- Verify that the blobs from the previous step are present in the subscribed namespace?
- Verify that the blobs from the previous step are present and complete in the subscribed namespace? In other words, there are no other blobs in the subscribed namespace.
- Verify the blobs from the previous step are included in an NMT with root `RowRoot`? Additionally verify a `RowRoot` is included in some `DataRoot`?

File: cips/cip-blob-verified-signer.md
Comment (rootulp) [inline cips/cip-blob-verified-signer.md:88]: I don't follow this security consideration. 

> Note this is a more secure assumption than forking which may require up to 2/3+ of the voting power to be "correct".

Does "forking" in this context apply to Celestia or the roll-up? Why do 2/3 of the voting power need to be correct to fork Celestia? Hypothetically if 2/3 voting power on the Celestia consensus layer are malicious, the 1/3 could create a fork that excludes the 2/3 malicious nodes.

> Rollups may decide to still retrieve the PFB's and validate the signatures themselves if they wish to avoid this assumption.

I think this is trying to say that regardless of the number of "incorrect" voting power on the DA layer, the DA consensus layer can't include a PFB with an signature that would convince an honest roll-up node to be deceived. But I don't really understand how verifying the PFB signatures avoids an assumption from above about 1/3 or 2/3 correct voting power.

File: cips/cip-blob-verified-signer.md
Comment (rootulp) [inline cips/cip-blob-verified-signer.md:2]: [optional] this title is a bit unclear because blobs are already authored by blob submitters. An alternative title that may be a bit more self-evident is: "Include signer in share format"

**Review by rootulp — COMMENTED** [review]
Thanks for including the diagram!

File: cips/cip-blob-verified-signer.md
Comment (cmwaters) [reply cips/cip-blob-verified-signer.md:2]: But blobs don't have the author / signer in the blob itself. It's embedded in the PFB. And this isn't just about changes to the share format

**Comment by cmwaters** [comment]
> Maybe we should add the fact that you need versioning for the CreateCommitment function. As we might be adding other things in the future which will change the shares and therefore the share commitment.
> 
> Create commitment will now have to be aware of the signer.

The CreateCommitment function has the following signature:

```
func CreateCommitment(blob *blob.Blob, merkleRootFn MerkleRootFn, subtreeRootThreshold int) ([]byte, error) 
```

As it includes the blob, which has the signer as well as the share version, I don't think anything will need to change to the function signature. Of course within the function, the commitment construction will depend on whether the signer is present in the blob or not

File: cips/cip-blob-verified-signer.md
Comment (cmwaters) [reply cips/cip-blob-verified-signer.md:88]: > Does "forking" in this context apply to Celestia or the roll-up? 

To Celestia (which also implies that there is a potential fork of all the rollups on Celestia).

What I was saying is that there is an assumption on the honesty of the validator set when using the signer (since the validators validate that the signer is correct) but that this assumption is in any way outweighed by the bigger assumption (i.e. the less secure attack vector) that there isn't currently a fork on Celestia

File: cips/cip-blob-verified-signer.md
Comment (cmwaters) [reply cips/cip-blob-verified-signer.md:69]: I would assume that inclusion and exclusion are done simultaneously i.e. we are proving that all the blobs we have are present in that namespace and there are no blobs in that namespace that we don't have.

Does that make sense? And would you like me to elaborate more in this document

File: cips/cip-blob-verified-signer.md
Comment (preston-evans98) [reply cips/cip-blob-verified-signer.md:72]: This data is already provable, since it's encoded in the start share of each blob anyway. But TBC, we at Soveregin don't actually need/want this - I was just pointing it out in case it's useful for optimistic rollups

File: cips/cip-blob-verified-signer.md
Comment (rootulp) [reply cips/cip-blob-verified-signer.md]: That sounds like the second option:

> Verify that the blobs from the previous step are present and complete in the subscribed namespace? In other words, there are no other blobs in the subscribed namespace.

But I think this would be clearer than what I previously wrote:

```
1. Retrieve all blobs in the subscribed namespace
2. Verify that the blobs from the previous step belong to the subscribed namespace and the list is complete (i.e. there are no blobs present in the subscribed namespace that are absent from the list of retrieved blobs).
```

File: cips/cip-blob-verified-signer.md
Comment (rootulp) [reply cips/cip-blob-verified-signer.md:90]: Ahh got it, thanks for expanding here.

File: cips/cip-blob-verified-signer.md
Comment (rootulp) [reply cips/cip-blob-verified-signer.md]: Good point. An alternative name could be: "Blob v2 with signer"

File: cips/cip-blob-verified-signer.md
Comment (mastergaurang94) [inline cips/cip-blob-verified-signer.md:71]: (Disclaimer: I'm a newb) what do you mean? Aren't the blobs returned guaranteed to be included?

File: cips/cip-blob-verified-signer.md
Comment (rootulp) [inline cips/cip-blob-verified-signer.md:62]: [nit] the current share version is `0` so a new share version would be `1`. To avoid confusion, we may want to replace v1 => 0 and v2 => 1 in these lines.

Ref: https://github.com/celestiaorg/celestia-app/blob/c5b0b9421e8bbe04d26104716bb86750a6891ab0/specs/src/specs/shares.md?plain=1#L49-L50

**Review by rootulp — APPROVED** [review]

File: cips/cip-blob-verified-signer.md
Comment (jcstein) [reply cips/cip-blob-verified-signer.md:62]: let's do this and I think we can get this (1) assigned a CIP number (2) edit the image directory and path in .md (3) and merged

File: cips/cip-blob-verified-signer.md
Comment (jcstein) [reply cips/cip-blob-verified-signer.md:33]: what do you think is best here @evan-forbes and @cmwaters, "in circuit" or "in zero-knowledge"?

File: cips/cip-blob-verified-signer.md
Comment (jcstein) [reply cips/cip-blob-verified-signer.md]: is this resolved?

File: cips/cip-blob-verified-signer.md
Comment (jcstein) [reply cips/cip-blob-verified-signer.md]: is this resolved?

File: cips/cip-blob-verified-signer.md
Comment (jcstein) [reply cips/cip-blob-verified-signer.md:74]: > `The location of the blob. Alternatively, we could put the blob commitment`
I think that this is not going to be the most atomic way to point to a blob. If I submit 2 blobs that are identical, their commitments will be the same, and their index will also be different, although they have the same commitment.

File: cips/cip-blob-verified-signer.md
Comment (jcstein) [reply cips/cip-blob-verified-signer.md:90]: is this resolved?

**Review by jcstein — CHANGES_REQUESTED** [review]
LGTM, but would like to resolve the open discussions before merging

File: cips/cip-blob-verified-signer.md
Comment (rootulp) [reply cips/cip-blob-verified-signer.md:90]: Yea resolved.

File: cips/cip-blob-verified-signer.md
Comment (cmwaters) [reply cips/cip-blob-verified-signer.md]: Yes this is resolved

File: cips/cip-blob-verified-signer.md
Comment (cmwaters) [reply cips/cip-blob-verified-signer.md:62]: Yes ok, I can also refer to them as v0 and v1

File: cips/cip-blob-verified-signer.md
Comment (cmwaters) [reply cips/cip-blob-verified-signer.md]: Resolved: https://github.com/celestiaorg/CIPs/pull/147/commits/62e5ef1ac8ffb785e5777cd837439d24b8a201f6

File: cips/cip-blob-verified-signer.md
Comment (cmwaters) [reply cips/cip-blob-verified-signer.md]: Changed to in circuit (https://github.com/celestiaorg/CIPs/pull/147/commits/62e5ef1ac8ffb785e5777cd837439d24b8a201f6)

File: cips/cip-blob-verified-signer.md
Comment (cmwaters) [reply cips/cip-blob-verified-signer.md]: Resolved in https://github.com/celestiaorg/CIPs/pull/147/commits/62e5ef1ac8ffb785e5777cd837439d24b8a201f6

**Comment by cmwaters** [comment]
@jcstein all comments should be resolved now

File: cips/cip-blob-verified-signer.md
Comment (cmwaters) [reply cips/cip-blob-verified-signer.md:58]: Yeah could the CIP maintainers update the file name and reference when they assign the CIP number

**Comment by rootulp** [comment]
Assigned this CIP-21. FYI @jcstein

File: cips/cip-blob-verified-signer.md
Comment (rootulp) [reply cips/cip-blob-verified-signer.md]: Done in https://github.com/celestiaorg/CIPs/pull/147/commits/0876f5bef4e04aad5770f2102fae29b8f2d06181

File: cips/cip-21.md
Comment (jcstein) [inline cips/cip-21.md:1]:
```suggestion
---
cip: 21
```

**Review by jcstein — APPROVED** [review]

**Pull request merged**
